third_party/libunwindstack/MapInfo.cpp

```diff
@@ -176,6 +192,146 @@ Memory* MapInfo::GetFileMemory() {
   return nullptr;
 }
 
+// Unlike loadable sections of ELF files, the .text section (and other sections) of a Portable
+// Executable can have an offset in the file (PointerToRawData, multiple of FileAlignment) that is
+// not congruent to the offset of that section when loaded into memory (VirtualAddress, multiple of
+// SectionAlignment) modulo the page size. This doesn't fulfill the requirements on the arguments of
+// mmap, so in these cases Wine cannot create a file-backed mapping for the .text section, and
+// resorts to creating an anonymous mapping and copying the .text section into it. This means that,
+// for PE binaries with this property, we cannot simply associate an executable mapping to the
+// corresponding file using the path in the mapping.
+//
+// However, we can make an educated guess. The path of a PE will at least appear in the read-only
+// mapping that corresponds to the beginning of the file, which contains the headers (because the
+// offset in the file is zero and the address chosen for this mapping should always be a multiple of
+// the page size). Therefore, let's consider the file of the last file mapping that precedes this
+// anonymous mapping. If this file is a PE, and an executable file mapping for the .text section of
+// this PE is not present, and finally the offset and the size of this anonymous mapping are
+// compatible with the address range where the .text section would be loaded based on the header for
+// the section (in particular, VirtualAddress and VirtualSize), then we can be quite sure that this
+// anonymous mapping indeed contains the .text section of that PE. Note that we assume that a PE
+// only has one .text section and one executable mapping.
+//
+// This method is the counterpart of GetFileMemory for anonymous executable mappings and contains
+// the detection mechanism described.
+Memory* MapInfo::GetFileMemoryFromAnonExecMapIfPeCoffTextSection() {
+  if (!name().empty()) {
+    // The mapping already has a name.
+    return nullptr;
+  }
+  if ((flags() & PROT_EXEC) == 0) {
+    // The mapping is not executable.
+    return nullptr;
+  }
+
+  std::shared_ptr<MapInfo> map_info_it = prev_map();
+
+  // Note that when the first character of a map name is '[', the mapping is a special one like
+  // [stack], [heap], etc. Even if such mapping has a name, it's not file-backed. An alternative way
+  // of detecting that would be to check if the inode is 0 (but that's not available here).
+  while (map_info_it != nullptr &&
```

Comment:
I don't get the case with the special mappings.
So I see two cases:
1. I am currently in an anonymous executable mapping and look-behind through the anonymous mappings of the same file, until I find a mapping with a file name. This one is at offset 0, so the alignment matches and we have the file name of our executable anonymous mapping - fine.
2. I am at again at an anonymous executable mapping and look-behind. Now, before I see a file-named mapping, I see one of the special mappings. Why can I simply skip this mapping? Can those mappings appear somewhere in the middle of a file mapped anonymously?

Sorry, if I miss something obvious here.

Reply:
> Can those mappings appear somewhere in the middle of a file mapped anonymously?

This is a very good question. I thought about this a bit when writing this and this was my conclusion:

First of all, I really think they can't.

However, conceptually, here I have an anonymous mapping and I'm looking for the previous file mapping. I identify a file mapping by it having a name that doesn't start with '['. That is all. Without considering '[', a special mapping could be considered a file mapping.

Now, we could say that if the mapping found (file or special) is special instead of file, we can return. I think that this adds one special case that also reduces generality without really providing a benefit.

Reply:
Thanks for the input. If you are confident that this does not happen, I am fine with keeping it the way it is. Otherwise, it looks that we could simplify the code a bit. Don't have a strong opinion. Maybe add a comment.

Reply:
If I understand this correctly, then my question here is if we should consider the case of a special mapping ("[...]") in the middle of a our search an error case. 

It seems the code would allow a case such case:

1) mapping: file.dll (header)
2) mapping: [special]
3) mapping: (anonymous and executable)

That said, if this is indeed wrong, we will for sure find out later when we check offsets and where the .text section of file.dll is mapped.

Reply:
SGTM, thanks for the additional example